### PR TITLE
AuditLogMethods

### LINKED ISSUE #34: Audit log methods.
General ideas include:

- `MessagePinEntryInfo#fetch_channel`
- `MessagePinEntryInfo#fetch_message`
- `MessageDeleteEntryInfo#fetch_channel`
- `MemberMoveEntryInfo#fetch_channel`
- `AuditLogEntry#fetch_user`

These should be tested and documented.

### PR DESCRIPTION
**Pull request by Forbidden-A**
### Checklist

- [x] I have run `nox` and all the pipelines have passed.
- [ ] I have made unittests according to the code I have added/modified/deleted.

### Related issues
Closes #34

### CODE REVIEW

File: hikari/audit_logs.py [hunk 1]
```diff
@@ -47,9 +47,14 @@
 import attr
 
 from hikari import snowflakes
+from hikari.channels import GuildTextChannel
```
Comment (davfsa) [inline hikari/audit_logs.py:50]: We rather not have singular imports as it is not the code style we are going for.

These should be changed to `from hikari import channels` and then access it using `channels.GuildTextChannel`. Also, if these are for typehinting, please move them under `if typing.TYPE_CHECKING`

Edit:
I just realized `from hikari import channels` is already under `if typing.TYPE_CHECKING`, so you can just remove these and use what I said above

Comment (Forbidden-A) [reply hikari/audit_logs.py:50]: aight

File: hikari/audit_logs.py [hunk 2]
```diff
@@ -216,6 +224,48 @@ class MessagePinEntryInfo(BaseAuditLogEntryInfo):
     message_id: snowflakes.Snowflake = attr.ib(repr=True)
     """The ID of the message that's being pinned or unpinned."""
 
+    async def fetch_channel(self) -> TextChannel:
+        """Fetch the object of the channel where a pinned message is being targeted.
+
+        Returns
+        -------
+        hikari.channels.TextChannel
```
Comment (davfsa) [inline hikari/audit_logs.py:232]: Short description missing here. You can use something like this:
```
The requested text channel.
```

File: hikari/audit_logs.py (hunk 1 above)
Comment (Forbidden-A) [reply hikari/audit_logs.py:50]: but I also use them in `assert isinstance(channel, TextChannel)`

File: hikari/events/__init__.pyi [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+from hikari.events.lifetime_events import *
+from hikari.events.member_events import *
+from hikari.events.message_events import *
+from hikari.events.reaction_events import *
+from hikari.events.role_events import *
+from hikari.events.shard_events import *
+from hikari.events.typing_events import *
+from hikari.events.user_events import *
+from hikari.events.voice_events import *
+from hikari.events.base_events import Event as Event, ExceptionEvent as ExceptionEvent
```
Comment (davfsa) [inline hikari/events/__init__.pyi:12]: How did you manage to add these :sweat_smile:. I thought they were in the `.gitignore`. Could you please remove them?

File: hikari/audit_logs.py [hunk 4]
```diff
@@ -260,7 +332,29 @@ class MemberMoveEntryInfo(MemberDisconnectEntryInfo):
...
+
+        Raises
+        ------
+        hikari.errors.UnauthorizedError
+            If you are unauthorized to make the request (invalid/missing token).
+        hikari.errors.ForbiddenError
+            If you are missing the `READ_MESSAGES` permission in the channel.
+        hikari.errors.NotFoundError
+            If the channel is not found.
+        hikari.errors.InternalServerError
```
Comment (davfsa) [inline hikari/audit_logs.py:352]: `RateLimitTooLongError` and  `RateLimitedError` missing here

**Review by davfsa — CHANGES_REQUESTED** [review]
Thank you for taking the time to work on this, just some little comments :+1:

File: hikari/audit_logs.py (hunk 1 above)
Comment (davfsa) [reply hikari/audit_logs.py:50]: In this cause you need to move `from hikari import channels` out of `if typing.TYPE_CHECKING:` and into the root and then use `channels.TextChannel`

File: hikari/events/__init__.pyi (hunk 3 above)
Comment (Forbidden-A) [reply hikari/events/__init__.pyi:12]: lmao

File: hikari/audit_logs.py (hunk 4 above)
Comment (Forbidden-A) [reply hikari/audit_logs.py:352]: right

File: hikari/__init__.pyi [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+from hikari.webhooks import *
+from hikari._about import __author__ as __author__, __ci__ as __ci__, __copyright__ as __copyright__, __discord_invite__ as __discord_invite__, __docs__ as __docs__, __email__ as __email__, __issue_tracker__ as __issue_tracker__, __license__ as __license__, __url__ as __url__, __version__ as __version__
+from hikari.files import Bytes as Bytes, File as File, LazyByteIteratorish as LazyByteIteratorish, Pathish as Pathish, Rawish as Rawish, Resourceish as Resourceish, URL as URL
+from hikari.impl.bot import BotApp as BotApp
+from hikari.impl.rest import RESTApp as RESTApp
+from hikari.snowflakes import SearchableSnowflakeish as SearchableSnowflakeish, SearchableSnowflakeishOr as SearchableSnowflakeishOr, Snowflake as Snowflake, Snowflakeish as Snowflakeish, SnowflakeishOr as SnowflakeishOr, Unique as Unique
+from hikari.undefined import UNDEFINED as UNDEFINED, UndefinedNoneOr as UndefinedNoneOr, UndefinedOr as UndefinedOr, UndefinedType as UndefinedType
+from typing import Any
+
+__all__: Any
```
Comment (davfsa) [inline hikari/__init__.pyi:31]: Remove this one too, since its generated when uploading the package to pypi

File: hikari/audit_logs.py [hunk 6]
```diff
@@ -243,7 +293,29 @@ class MessageDeleteEntryInfo(MessageBulkDeleteEntryInfo):
...
```
Comment (davfsa) [inline hikari/audit_logs.py:313]: `RateLimitTooLongError` and `RateLimitedError` also missing here

File: hikari/audit_logs.py [hunk 7]
```diff
@@ -216,6 +224,48 @@ class MessagePinEntryInfo(BaseAuditLogEntryInfo):
...
+        Returns
+        -------
+        hikari.messages.Message
+
+        Raises
+        ------
+        hikari.errors.UnauthorizedError
+            If you are unauthorized to make the request (invalid/missing token).
+        hikari.errors.ForbiddenError
+            If you are missing the `READ_MESSAGES` permission in the channel that the message is in.
```
Comment (davfsa) [inline hikari/audit_logs.py:261]: And here :sweat_smile:

File: hikari/audit_logs.py [hunk 8]
```diff
@@ -310,6 +445,38 @@ class AuditLogEntry(snowflakes.Unique):
     reason: typing.Optional[str] = attr.ib(eq=False, hash=False, repr=False)
     """The reason for this change, if set (between 0-512 characters)."""
 
+    async def fetch_user(self) -> typing.Optional[users_.User]:
+        """Fetch the user who made this change.
+
+        Returns
+        -------
+        typing.Optional[hikari.users.user]
+            The requested user
```
Comment (davfsa) [inline hikari/audit_logs.py:454]: This should be `The user who made this change, if available`

File: hikari/audit_logs.py [hunk 9]
```diff
@@ -260,7 +362,40 @@ class MemberMoveEntryInfo(MemberDisconnectEntryInfo):
...
-    """The amount of members who were disconnected from voice in this entry."""
+    """The channel that the member(s) have been moved to"""
+
+    async def fetch_channel(self) -> channels.GuildVoiceChannel:
+        """Fetch the guild text based channel where these message(s) were deleted.
+
+        Returns
+        -------
+        hikari.channels.GuildVoiceChannel
+
```
Comment (davfsa) [inline hikari/audit_logs.py:373]: Forgot to mention short description for the return is also missing here

File: hikari/audit_logs.py [hunk 10]
```diff
@@ -243,7 +312,40 @@ class MessageDeleteEntryInfo(MessageBulkDeleteEntryInfo):
...
-    """The guild text based channel where these message(s) were deleted."""
+    """The ID of guild text based channel where these message(s) were deleted."""
+
+    async def fetch_channel(self) -> channels.GuildTextChannel:
+        """Fetch the guild text based channel where these message(s) were deleted.
+
+        Returns
+        -------
+        hikari.channels.GuildTextChannel
+
```
Comment (davfsa) [inline hikari/audit_logs.py:323]: And here

File: hikari/audit_logs.py [hunk 11]
```diff
@@ -216,6 +220,71 @@ class MessagePinEntryInfo(BaseAuditLogEntryInfo):
...
+        assert isinstance(channel, channels.TextChannel)
+        return channel
+
+    async def fetch_message(self) -> messages.Message:
+        """Fetch the object of the message that's being pinned or unpinned.
+
+        Returns
+        -------
+        hikari.messages.Message
+
```
Comment (davfsa) [inline hikari/audit_logs.py:263]: And here

File: hikari/audit_logs.py [hunk 12]
```diff
@@ -310,6 +445,38 @@ class AuditLogEntry(snowflakes.Unique):
     reason: typing.Optional[str] = attr.ib(eq=False, hash=False, repr=False)
     """The reason for this change, if set (between 0-512 characters)."""
 
+    async def fetch_user(self) -> typing.Optional[users_.User]:
```
Comment (davfsa) [inline hikari/audit_logs.py:448]: I was just thinking about this, but would a `fetch_member` be more useful here? Maybe even as another method

Comment (Forbidden-A) [reply hikari/audit_logs.py:448]: there is not guild_id though

**Comment by Forbidden-A** [comment]
@davfsa I haven't written tests

**Pull request closed without merging**